src/Http/Http.Abstractions/src/Internal/PathStringHelper.cs

```diff
@@ -24,24 +27,57 @@ internal class PathStringHelper
             true,  true,  true,  false, false, false, true,  false,     // 0x78 - 0x7F
         };
 
+        // Map from an ASCII char to its hex value, e.g. arr['b'] == 11. 0xFF means it's not a hex digit.
+        // From coreclr/src/System.Private.CoreLib/shared/System/Number.Parsing.cs
+        // byte[] rather than ReadOnlySpan<byte> to avoid double .Length check https://github.com/dotnet/coreclr/issues/24209
+        private static readonly byte[] CharToHexLookup = new byte[]
+        {
+            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 15
+            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 31
+            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 47
+            0x0,  0x1,  0x2,  0x3,  0x4,  0x5,  0x6,  0x7,  0x8,  0x9,  0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 63
+            0xFF, 0xA,  0xB,  0xC,  0xD,  0xE,  0xF,  0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 79
+            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // 95
+            0xFF, 0xa,  0xb,  0xc,  0xd,  0xe,  0xf // 102
+        };
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool IsValidPathChar(char c)
         {
-            return c < ValidPathChars.Length && ValidPathChars[c];
+            // Use local array and uint .Length compare to elide the bounds check on array access
+            var validChars = ValidPathChars;
+            var i = (int)c;
+            return (uint)i < (uint)validChars.Length && validChars[i];
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static bool IsPercentEncodedChar(string str, int index)
         {
-            return index < str.Length - 2
-                && str[index] == '%'
-                && IsHexadecimalChar(str[index + 1])
-                && IsHexadecimalChar(str[index + 2]);
+            if (str[index] == '%' && index < str.Length - 2)
```

Comment:
```suggestion
            var len = (uint)str.Length;
            if (str[index] == '%' && index < len - 2)
```
Reuses the str-length, instead of reloading.
```diff
G_M56184_IG02:
+	   8B4708               mov      eax, dword ptr [rdi+8]
+      8BD0                 mov      edx, eax
+      3BF0                 cmp      esi, eax
-      3B7708               cmp      esi, dword ptr [rdi+8]
       7320                 jae      SHORT G_M56184_IG05
       4863C6               movsxd   rax, esi
       66837C470C25         cmp      word  ptr [rdi+2*rax+12], 37
       7511                 jne      SHORT G_M56184_IG03
+      4863C6               movsxd   rax, esi	   
-      8B4708               mov      eax, dword ptr [rdi+8]
       83C0FE               add      eax, -2
+      483BC2               cmp      rax, rdx	   
-      3BC6                 cmp      eax, esi
       7E07                 jle      SHORT G_M56184_IG03
       E824F4FFFF           call     PathStringHelper:AreFollowingTwoCharsHex(ref,int):bool
       EB02                 jmp      SHORT G_M56184_IG04
```